### PR TITLE
[Serve] Java documentation

### LINKED ISSUE #27580: [serve][docs] Basic Java Documentation
In Ray 2.0, Serve has added the ability to serve Java replicas as well as the full Java client API. We should add documentation to showcase the features.

### PR DESCRIPTION
**Pull request by liuyang-my**
## Why are these changes needed?

Now we have commit Ray Serve Java API. Users can use Java to deploy and access Ray Serve Deployment end-to-end. But there is no relevant documents. This PR will commit the documents about Ray Serve Java user case.

## Related issue number

#22726

Closes #27580

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
Great start. Here's my recommendation:

## Location
- move the getting started page as Serve's example. Put it in [here](https://github.com/ray-project/ray/tree/master/doc/source/serve/tutorials) and add an entry to [this page](https://github.com/ray-project/ray/blob/master/doc/source/serve/tutorials/index.md)
- keep the existing managing deployment page but reference it in user guide: https://github.com/ray-project/ray/blob/master/doc/source/_toc.yml#L159-L171

## Content: `doc/source/serve/getting_started_java.md`
This page should be converted to end to end tutorial runnable by user. Imagine user just copy and past and it should work. 
The general structure make sense; please add a page about using handle to call.

## Content: `doc/source/serve/managing-deployments-java.md`
This page should cover more "how do I do X using the API". Because this is new, we should also discuss
- Why should user consider this?
- What's the benefit of using the Java API
- Cross Language API
- Future roadmap

File: doc/source/serve/managing-java-deployments.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
+# Managing Java Deployments
+
+Java is one of the mainstream programming languages. Some users are used to using Java for daily development. Therefore, we provide the Ray Serve Java API. Through this Java API, users can manage the deployments written in Java. At the same time, Java API can also manage Python model deployments across languages.
```
Comment (simon-mo) [inline doc/source/serve/managing-java-deployments.md:3]:
```suggestion
Java is one of the mainstream programming languages for production services. Ray Serve natively supports Java API for creating, updating and managing deployments. You can create deployment using Java and call them via Python, or vice versa. 
```

File: doc/source/serve/managing-java-deployments.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+This section should help you:
+
+- create, query, update and configure Java deployments
+- configure resources of your Java deployments
+- manage Python deployments
```
Comment (simon-mo) [inline doc/source/serve/managing-java-deployments.md:9]:
```suggestion
- manage Python deployments using Java API
```

File: doc/source/serve/managing-java-deployments.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+```
+
+> NOTE: Before `Ray.init` or `Serve.start`, we need to set the directory to find the Python code. For details, please refer to [Cross-Language Programming](cross_language).
+
+## Future Roadmap
+
+In the future, we will provide more features on Ray Serve Java, such as:
+- better Java API experience
+- HTTP ingress for Ray Serve Java
+- deploy a Java Spring project as a deployment
```
Comment (simon-mo) [inline doc/source/serve/managing-java-deployments.md:192]:
```suggestion
- improved API to match the Python version
- HTTP ingress support
- bring your own Java Spring project as a deployment
```

File: doc/source/serve/tutorials/java.md [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+```xml
+<dependency>
+  <groupId>io.ray</groupId>
+  <artifactId>ray-serve</artifactId>
+  <version>${ray.version}</version>
+  <scope>provided</scope>
+</dependency>
+```
+
+> NOTE: After installing Ray, the Java jar of Ray Serve has been included locally. The `provided` scope could  ensure the Java code using Ray Serve can be compiled and will not cause version conflicts when deployed on the cluster.
```
Comment (simon-mo) [inline doc/source/serve/tutorials/java.md:16]:
```suggestion
> NOTE: After installing Ray, the Java jar of Ray Serve has been included locally. The `provided` scope could ensure the Java code using Ray Serve can be compiled and will not cause version conflicts when deployed on the cluster.
```

File: doc/source/serve/tutorials/java.md [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+
+## Example Model
+
+This is a real user scenario we encountered. The user's class is as follows:
```
Comment (simon-mo) [inline doc/source/serve/tutorials/java.md:20]:
```suggestion
Our example use case is derived from production workflow of a financial application. The application needs to compute the best strategy to interact with different banks for a single task. 
```

**Review by simon-mo — COMMENTED** [review]
Seems like the code is duplicated. You can use the literalinclude directive to embed the file without copying them https://myst-parser.readthedocs.io/en/latest/faq/index.html?highlight=literalinclude#include-a-file-from-outside-the-docs-folder-like-readme-md

File: doc/source/serve/managing-java-deployments.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
+# Managing Java Deployments
+
+Java is one of the mainstream programming languages for production services. Ray Serve natively supports Java API for creating, updating and managing deployments. You can create deployment using Java and call them via Python, or vice versa.
+
```
Comment (dmatrix) [inline doc/source/serve/managing-java-deployments.md:4]: Java is one of the mainstream programming languages for production services. Ray Serve natively supports Java API for creating, updating, and managing deployments. You can create serve deployment using ....

Comment (dmatrix) [inline doc/source/serve/managing-java-deployments.md:4]: updating, and managing deployments..

Comment (dmatrix) [reply doc/source/serve/managing-java-deployments.md:4]: Might want to use serial comma (https://academicguides.waldenu.edu/writingcenter/apa/other/serialcommas)

Comment (liuyang-my) [reply doc/source/serve/managing-java-deployments.md:4]: Thank you for your suggestion. It has been modified according to the `Serial Comma`.

**Comment by simon-mo** [comment]
@dmatrix can you take another look? here's the rendering of the two added pages:
https://ray--26321.org.readthedocs.build/en/26321/serve/managing-java-deployments.html
https://ray--26321.org.readthedocs.build/en/26321/serve/tutorials/java.html

File: doc/source/serve/tutorials/java.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
+(serve-java-tutorial)=
+
+# Java Tutorial
+
+First of all, using Java Ray Serve needs the following dependency in you `pom.xml`:
+
```
Comment (dmatrix) [inline doc/source/serve/tutorials/java.md:6]: To use Java Ray Serve, you need the following dependency in your `pom.xml.`

File: doc/source/serve/tutorials/java.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+<dependency>
+  <groupId>io.ray</groupId>
+  <artifactId>ray-serve</artifactId>
+  <version>${ray.version}</version>
+  <scope>provided</scope>
+</dependency>
+```
+
+> NOTE: After installing Ray, the Java jar of Ray Serve has been included locally. The `provided` scope could ensure the Java code using Ray Serve can be compiled and will not cause version conflicts when deployed on the cluster.
+
```
Comment (dmatrix) [inline doc/source/serve/tutorials/java.md:17]: After installing Ray, the Java jar of Ray Serve is included locally.

File: doc/source/serve/tutorials/java.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+Next, we start the Ray Serve runtime and deploy `StrategyOnRayServe` as a deployment.
+
+```{literalinclude} ../../../../java/serve/src/test/java/io/ray/serve/docdemo/StrategyCalcOnRayServe.java
+:end-before: docs-deploy-end
+:language: java
+:start-after: docs-deploy-start
+```
+
+The `Deployment.create` makes a `Deployment` object named "strategy". After executing `Deployment.deploy`, this "strategy" deployment is deployed in the instance of Ray Serve with 4 replicas, and we can access it for distributed parallel computing.
```
Comment (dmatrix) [inline doc/source/serve/tutorials/java.md:65]: The `Deployment.create` makes a `Deployment` object named "strategy."

Comment (dmatrix) [inline doc/source/serve/tutorials/java.md:65]: 4 replicas --> four replicas

File: doc/source/serve/tutorials/java.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+Now we can test the "strategy" deployment using RayServeHandle inside Ray:
+
+```{literalinclude} ../../../../java/serve/src/test/java/io/ray/serve/docdemo/StrategyCalcOnRayServe.java
+:end-before: docs-calc-end
+:language: java
+:start-after: docs-calc-start
+```
+
+At present, the calculation of each bank's each indicator is still in series, and just sended to Ray for execution. We can make the calculation concurrent, which not only improves the calculation efficiency, but also solves the bottleneck of single machine.
+
```
Comment (dmatrix) [inline doc/source/serve/tutorials/java.md:78]: what does "is still in series," mean? Might want to clarify. Do you mean serial?

File: doc/source/serve/tutorials/java.md [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+```{literalinclude} ../../../../java/serve/src/test/java/io/ray/serve/docdemo/StrategyCalcOnRayServe.java
+:end-before: docs-main-end
+:language: java
+:start-after: docs-main-start
+```
+
+## Calling Ray Serve Deployment with HTTP
+
+Another way to call a deployment is through the HTTP request. But there are now two limitations for the Java deployments:
+
```
Comment (dmatrix) [inline doc/source/serve/tutorials/java.md:96]: Another way to test or call a deployment is ...

File: doc/source/serve/tutorials/java.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+This code can also be rewritten to support concurrency:
+
+```{literalinclude} ../../../../java/serve/src/test/java/io/ray/serve/docdemo/HttpStrategyCalcOnRayServe.java
+:end-before: docs-parallel-calc-end
+:language: java
+:start-after: docs-parallel-calc-start
+```
+
+Now, the complete usage of `HttpStrategyCalcOnRayServe` is like this:
+
```
Comment (dmatrix) [inline doc/source/serve/tutorials/java.md:140]: Finally, the complete usage ....

**Review by dmatrix — CHANGES_REQUESTED** [review]

**Comment by simon-mo** [comment]
@dmatrix addressed!

File: doc/source/serve/tutorials/java.md (hunk 10 above)
Comment (dmatrix) [reply doc/source/serve/tutorials/java.md:78]: just sended to Ray --> sent to Ray

**Review by dmatrix — APPROVED** [review]
suggested minor nit otherwise LGTM

**Comment by simon-mo** [comment]
Need Java code owner here as well. @WangTaoTheTonic, @kfstorm or @jovany-wang can you please approve?

**Review by jovany-wang — APPROVED** [review]

**Pull request merged**